README.md
# 🎶 @tmigone/pulseaudio 🎶
 ![NPM](https://img.shields.io/npm/v/@tmigone/pulseaudio.svg?logo=npm&logoColor=fff&label=NPM+package&color=limegreen&link=https://www.npmjs.com/package/@tmigone/pulseaudio) ![CI](https://github.com/tmigone/pulseaudio/workflows/ci/badge.svg) [![Coverage Status](https://coveralls.io/repos/github/tmigone/pulseaudio/badge.svg?branch=refactor)](https://coveralls.io/github/tmigone/pulseaudio?branch=refactor)

`@tmigone/pulseaudio` is a TypeScript based client library for [PulseAudio](https://www.freedesktop.org/wiki/Software/PulseAudio/), the most popular sound server for Linux. This library allows you to easily build clients or applications that interact with a PulseAudio server over it's native protocol, for example media players/recorders, volume control applications, etc.


### Features
- Zero dependency fully typed TypeScript implementation of the PulseAudio client protocol
- Extensive testing suite
- Protocol features: 
  - authentication - provide authentication data for the server
  - transport - connect over UNIX domain sockets or TCP sockets
  - introspection - query, modify and operate on PulseAudio objects like modules, sinks, sources, etc.
  - events - subscribe to server-side object events like a sink starting playback, etc.
  - (To be implemented) streams - manage audio playback and recording using Node.js streams

## Installation

Install the library using [npm](https://www.npmjs.com/):
```bash
npm install @tmigone/pulseaudio
```

## Usage

```ts
import PulseAudio, { Sink } from '@tmigone/pulseaudio'

(async () => {
  // Connect using tcp or unix socket
  // const client: PulseAudio = new PulseAudio('unix:/run/pulse/pulseaudio.socket')
  const client: PulseAudio = new PulseAudio('tcp:192.168.1.10:4317')
  await client.connect()

  // Set volume of all sinks to 50%
  const sinks: Sink[] = await client.getSinkList()
  for (const sink of sinks) {
    await client.setSinkVolume(sink.index, 50)
  }

  // Close connection
  client.disconnect()
})()
```

## Documentation

Visit the [docs site](http://pulseaudio.tmigone.com) for in depth documentation on the library API's.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.5 - 2022-08-25

* client: correctly wrap some lines [SpacingBat3]
* index: export `UnixSocket` interface [SpacingBat3]
* meta: add myself to the contributors list [SpacingBat3]
* client,tests: add support for `unix:` addresses. [SpacingBat3]

## 1.0.4 - 2022-01-26

* commands: fix return type of several commands [Tomás Migone]

## 1.0.3 - 2022-01-26

* command: fix getSourceList return type [Tomás Migone]

## 1.0.2 - 2022-01-25

* client: remove timeout callback [Tomás Migone]

## 1.0.1 - 2022-01-20

* meta: fix workflows and readme typos [Tomás Migone]

## 1.0.0 - 2022-01-20

* docs: add typedoc generated documentation site [Tomás Migone]
* tests: add more tests [Tomás Migone]
* tests: add missing unloadModule tests [Tomás Migone]
* workflows: fix coverage action [Tomás Migone]
* docs: add basic readme, add code coverage boilerplate [Tomás Migone]
* Rename PAClient to PulseAudio [Tomás Migone]
* Fix channelMap and channelVolume bug [Tomás Migone]
* Fix ci [Tomás Migone]
* lint: add lint tooling [Tomás Migone]
* test: separate unit from e2e, fix ci [Tomás Migone]
* e2e: add first basic test [Tomás Migone]
* source-output: add commands and tests [Tomás Migone]
* source: add source commands and tests [Tomás Migone]
* command: add module commands [Tomás Migone]
* paprop: fix bug where property list were parsed incorrectly [Tomás Migone]
* commands: add server commands with tests [Tomás Migone]
* command: refactor client commands, add tests [Tomás Migone]
* command: add sinkInput commands and tests [Tomás Migone]
* sink: rename getSinks to getSinkList to follow PA naming scheme [Tomás Migone]
* docs: add testing readme [Tomás Migone]
* tests: refactor testing framework, add sink tests [Tomás Migone]
* command: refactor sink commands [Tomás Migone]
* Multiple changes: - refactor e2e test boilerplate - add iterator - fix sink commands [Tomás Migone]
* client: init implementation of protocol versions with sink command rework [Tomás Migone]
* tests: add setup for e2e tests and setup instructions [Tomás Migone]
* client: bump client protocol version to v35 (pa v15) [Tomás Migone]
* client: add disconnect method [Tomás Migone]

## 0.1.12 - 2020-12-24

* meta: fix deploy workflow [Tomás Migone]
* meta: use admin token for workflows [Tomás Migone]
* test [Tomás Migone]

## 0.1.11 - 2020-12-24

* meta: fix deploy workflow [Tomás Migone]
* meta: use admin token for workflows [Tomás Migone]
* test [Tomás Migone]
* meta: fix deploy workflow [Tomás Migone]
* meta: test actions [Tomás Migone]
* meta: fix deploy workflow [Tomás Migone]
* meta: add nodejs 10.x test to CI [Tomás Migone]
* meta: refactor GitHub CI/CD workflows [Tomás Migone]

## 0.1.10 - 2020-12-20

* Remove VERSION file [Tomás Migone]
* versionist: update repo.yml type [Tomás Migone]

# v0.1.9
## (2020-12-20)

* meta: fix versionist action [Tomás Migone]
* meta: Update CD action [Tomás Migone]

# v0.1.8
## (2020-08-18)

* client: Make connect public [Tomás Migone]

# v0.1.7
## (2020-08-17)

* client: add moveSinkInput and getSinkInputs commands [Tomás Migone]

# v0.1.6
## (2020-08-12)

* Allow auth without cookie [Tomás Migone]

# v0.1.5
## (2020-08-10)

* Add u8 tag [Tomás Migone]

# v0.1.4
## (2020-08-07)

* Fix PAPacket bug that resulted in lost events [Tomás Migone]

# v0.1.3
## (2020-08-07)

* Fix Format tag [Tomás Migone]

# v0.1.2
## (2020-07-03)

* lib: remove lib test [Tomás Migone]

# v0.1.1
## (2020-07-03)

* export types through index [Tomás Migone]

# v0.1.0
## (2020-07-03)

* types: rework outward pulseaudio types [Tomás Migone]

# v0.0.12
## (2020-07-01)

* Changes: [Tomás Migone]

# v0.0.11
## (2020-07-01)

* changes: [Tomás Migone]

# v0.0.10
## (2020-06-30)

* commands: getSinks allow multiple sinks returned [Tomás Migone]
* proplist: fix bug where empty strings on proplist broke tag parsing [Tomás Migone]

# v0.0.9
## (2020-06-26)

* npm: publish all files in build/ [Tomás Migone]

# v0.0.8
## (2020-06-26)

* meta: rework npm scripts [Tomás Migone]

# v0.0.7
## (2020-06-26)

* npm: require node 12.4.x [Tomás Migone]

# v0.0.6
## (2020-06-26)

* Test [Tomás Migone]

# v0.0.5
## (2020-06-26)

* Test [Tomás Migone]

# v0.0.4
## (2020-06-26)

* Test [Tomás Migone]

# v0.0.3
## (2020-06-26)

* meta: Replace prepublish script with prepare [Tomás Migone]

# v0.0.2
## (2020-06-26)

* meta: add versionist [Tomás Migone]
* meta: update test command [Tomás Migone]


src/client.ts
import { EventEmitter } from 'events'
import { readFileSync } from 'fs'
import { Socket } from 'net'
import PAPacket from './packet'
import PARequest from './request'
import { PATag } from './tag'
import { PA_MAX_REQUEST_ID, PA_PROTOCOL_MINIMUM_VERSION } from './protocol'
import { PASubscriptionEventType, PAEvent } from './event'
import { PAError } from './error'
import {
  AuthInfo,
  ClientInfo,
  Module,
  ServerInfo,
  Sink,
  SinkInput,
  Source,
  SourceOutput,
  SubscribeInfo,
  VolumeInfo
} from './types/pulseaudio'

import { SetClientName } from './commands/client'
import { PA_NATIVE_COMMAND_NAMES } from './commands'
import { GetSink, GetSinkList, SetSinkVolume } from './commands/sink'
import { Authenticate, GetServerInfo, Subscribe } from './commands/server'
import { GetSinkInput, GetSinkInputList, MoveSinkInput } from './commands/sinkInput'
import { GetModule, GetModuleList, LoadModule, UnloadModule } from './commands/module'
import { GetSource, GetSourceList, SetSourceVolume } from './commands/source'
import { GetSourceOutput, GetSourceOutputList, MoveSourceOutput } from './commands/sourceOutput'

export interface TCPSocket {
  type: 'tcp'
  port: number
  host: string
}

export interface UnixSocket {
  type: 'unix'
  path: string
}

/**
 * Implements the PulseAudio client.
 *
 * @example
 * ```ts
 * import PulseAudio, { Sink } from '@tmigone/pulseaudio'
 *
 * (async () => {
 *   // Connect using tcp socket
 *   const client: PulseAudio = new PulseAudio('tcp:192.168.1.10:4317')
 *   await client.connect()
 *
 *   // Set volume of all sinks to 50%
 *   const sinks: Sink[] = await client.getSinkList()
 *   for (const sink of sinks) {
 *     await client.setSinkVolume(sink.index, 50)
 *   }
 *
 *   // Close connection
 *   client.disconnect()
 * })()
 * ```
 * @noInheritDoc
 */
export default class PulseAudio extends EventEmitter {
  /**
  * @category public
  */
  public address: TCPSocket | UnixSocket
  /**
  * @category public
  */
  public cookie: Buffer = Buffer.allocUnsafe(256)
  /**
  * @category public
  */
  public connected: boolean = false
  /**
  * @category public
  */
  public protocol: number = 0
  private socket: Socket
  private chunks: Buffer[] = []
  private requests: PARequest[] = []
  private lastRequestId: number = 0

  /**
  * Instantiate a PulseAudio client by providing the server address.
  * PulseAudio server defaults to the following:
  * - UNIX socket: `/run/user/<user_id>/pulse/native`
  * - TCP port: `4317`
  *
  * ```typescript
  * // Using UNIX socket
  * const client: PulseAudio = new PulseAudio('unix:/run/user/1001/pulse/native')
  *
  * // Using TCP socket
  * const client: PulseAudio = new PulseAudio('tcp:localhost:4317')
  * ```
  * @category PulseAudio
  * @param address The address of the PulseAudio server. Can be either the path to a UNIX domain socket or the network address of the server. For UNIX sockets prepend `unix:` to the path. For TCP sockets the format is `tcp:<host>:<port>`.
  * @param cookie The path to the cookie to use when authenticating with the server. Typically located at `~/.config/pulse/cookie`.
  */
  constructor (address: string, cookie?: string) {
    super()
    this.parseAddress(address)
    if (cookie !== undefined) this.parseCookie(cookie)
  }

  /**
  * Connects the client to the PulseAudio server. Won't retry if the server is not reachable or the connection attempt is refused. Times out after 5 seconds.
  * @category client
  */
  async connect (): Promise<AuthInfo> {
    return await new Promise<AuthInfo>((resolve, reject) => {
      this.socket = new Socket()
      if (this.address.type === 'tcp') {
        this.socket.connect(this.address.port, this.address.host)
      } else {
        this.socket.connect(this.address.path)
      }
      // eslint-disable-next-line @typescript-eslint/no-misused-promises
      this.socket.on('connect', async () => {
        this.connected = true

        // Authenticate client
        const reply: AuthInfo = await this.authenticate()
        this.protocol = reply.protocol
        if (this.address.type === 'tcp') {
          console.log(`Connected to PulseAudio at tcp://${this.address.host}:${this.address.port} using protocol v${this.protocol}`)
        } else {
          console.log(`Connected to PulseAudio at unix://${this.address.path} using protocol v${this.protocol}`)
        }

        if (reply.protocol < PA_PROTOCOL_MINIMUM_VERSION) {
          this.disconnect()
          reject(new Error(`Server protocol version is too low, please update to ${PA_PROTOCOL_MINIMUM_VERSION} or higher.`))
        }

        resolve(reply)
      })
      this.socket.on('readable', this.onReadable.bind(this))
      this.socket.on('error', reject)
    })
  }

  /**
  * Disconnect the client from the PulseAudio server. Waits for the server to close the connection.
  * @category client
  */
  disconnect (): void {
    this.socket.removeAllListeners()
    this.socket.end()
  }

  /**
  * Specify the client name passed to the server when connecting.
  * @category client
  * @param clientName The client name. Defaults to `paclient`.
  */
  async setClientName (clientName: string = 'paclient'): Promise<ClientInfo> {
    const query: PAPacket = SetClientName.query(this.requestId(), clientName)
    return await this.sendRequest(query)
  }

  private async authenticate (): Promise<AuthInfo> {
    const query: PAPacket = Authenticate.query(this.requestId(), this.cookie)
    return await this.sendRequest(query)
  }

  /**
  * Subscribe to PulseAudio server events.
  * @category server
  */
  async subscribe (): Promise<SubscribeInfo> {
    const query: PAPacket = Subscribe.query(this.requestId())
    return await this.sendRequest(query)
  }

  /**
  * Gets PulseAudio server information.
  * @category server
  */
  async getServerInfo (): Promise<ServerInfo> {
    const query: PAPacket = GetServerInfo.query(this.requestId())
    return await this.sendRequest(query)
  }

  /**
  * Gets details for the sink instance identified by the specified symbolic name or numeric index.
  * @category sink
  * @param sink The symbolic name or numerical index of the sink instance to get details for.
  */
  async getSink (sink: number | string): Promise<Sink> {
    const query: PAPacket = GetSink.query(this.requestId(), sink)
    return await this.sendRequest(query)
  }

  /**
  * Gets details for all sink instances.
  * @category sink
  */
  async getSinkList (): Promise<Sink[]> {
    const query: PAPacket = GetSinkList.query(this.requestId())
    return await this.sendRequest(query)
  }

  /**
  * Sets the volume of the specified sink (identified by its symbolic name or numerical index).
  * @category sink
  * @param sink The symbolic name or numerical index of the sink to set the volume of.
  * @param volume The volume to set the source to in percentage (0% - 100%).
  */
  async setSinkVolume (sink: number | string, volume: number): Promise<VolumeInfo> {
    const query: PAPacket = SetSinkVolume.query(this.requestId(), sink, { channels: 2, volumes: [volume, volume] })
    return await this.sendRequest(query)
  }

  /**
  * Gets details for all sink input instances.
  * @category sinkInput
  */
  async getSinkInputList (): Promise<Sink[]> {
    const query: PAPacket = GetSinkInputList.query(this.requestId())
    return await this.sendRequest(query)
  }

  /**
  * Gets details for the sink input instance identified by the specified symbolic name or numeric index.
  * @category sinkInput
  * @param sinkInput The symbolic name or numerical index of the sink input instance to get details for.
  */
  async getSinkInput (sinkInput: number | string): Promise<SinkInput> {
    const query: PAPacket = GetSinkInput.query(this.requestId(), sinkInput)
    return await this.sendRequest(query)
  }

  /**
  * Move the specified playback stream (identified by its numerical index) to the specified sink (identified by its numerical index).
  * @category sinkInput
  * @param sinkInput The numerical index of the playback stream to move.
  * @param sink The numerical index of the sink to move the playback stream to.
  */
  async moveSinkInput (sinkInput: number, sink: number): Promise<any> {
    const query: PAPacket = MoveSinkInput.query(this.requestId(), sinkInput, sink)
    return await this.sendRequest(query)
  }

  /**
  * Gets details for the source instance identified by the specified symbolic name or numeric index.
  * @category source
  * @param source The symbolic name or numerical index of the source instance to get details for.
  */
  async getSource (source: number | string): Promise<Source> {
    const query: PAPacket = GetSource.query(this.requestId(), source)
    return await this.sendRequest(query)
  }

  /**
  * Gets details for all source instances.
  * @category source
  */
  async getSourceList (): Promise<Source[]> {
    const query: PAPacket = GetSourceList.query(this.requestId())
    return await this.sendRequest(query)
  }

  /**
  * Sets the volume of the specified source (identified by its symbolic name or numerical index).
  * @category source
  * @param source The symbolic name or numerical index of the source to set the volume of.
  * @param volume The volume to set the source to in percentage (0% - 100%).
  */
  async setSourceVolume (source: number | string, volume: number): Promise<Source> {
    const query: PAPacket = SetSourceVolume.query(this.requestId(), source, { channels: 2, volumes: [volume, volume] })
    return await this.sendRequest(query)
  }

  /**
  * Gets details for the source output instance identified by the specified numeric index.
  * @category sourceOutput
  * @param sourceOutput The numerical index of the source output instance to get details for.
  */
  async getSourceOutput (sourceOutput: number | string): Promise<SourceOutput> {
    const query: PAPacket = GetSourceOutput.query(this.requestId(), sourceOutput)
    return await this.sendRequest(query)
  }

  /**
  * Gets details for all source output instances.
  * @category sourceOutput
  */
  async getSourceOutputList (): Promise<SourceOutput[]> {
    const query: PAPacket = GetSourceOutputList.query(this.requestId())
    return await this.sendRequest(query)
  }

  /**
  * Move the specified recording stream (identified by its numerical index) to the specified source (identified by its numerical index).
  * @category sourceOutput
  * @param sourceOutput The numerical index of the recording stream to move.
  * @param source The numerical index of the source to move the recording stream to.
  */
  async moveSourceOutput (sourceOutput: number, source: number): Promise<SourceOutput> {
    const query: PAPacket = MoveSourceOutput.query(this.requestId(), sourceOutput, source)
    return await this.sendRequest(query)
  }

  /**
  * Gets details for the module instance identified by the specified numeric index.
  * @category module
  * @param module The numerical index of the module instance to get details for.
  */
  async getModule (module: number): Promise<Module> {
    const query: PAPacket = GetModule.query(this.requestId(), module)
    return await this.sendRequest(query)
  }

  /**
  * Gets details for all loaded module instances.
  * @category module
  */
  async getModuleList (): Promise<Module[]> {
    const query: PAPacket = GetModuleList.query(this.requestId())
    return await this.sendRequest(query)
  }

  /**
  * Load the specified module with the specified arguments into the running sound server. Prints the numeric index of the module just loaded to STDOUT. You can use it to unload the module later.
   *
   * @example
   * ```typescript
   * const moduleIndex = await client.loadModule('module-loopback', 'source=alsa_output.dac.stereo-fallback.monitor sink=app-output')
   * console.log(moduleIndex) // moduleIndex: 27
   * ```
  * @category module
  * @param name The name of the module to load.
  * @param argument Space separated list of arguments to pass to the module.
  */
  async loadModule (name: string, argument: string): Promise<Module> {
    const query: PAPacket = LoadModule.query(this.requestId(), name, argument)
    return await this.sendRequest(query)
  }

  /**
  * Unload the module instance identified by the specified numeric index.
  * @category module
  * @param module The numeric index of the module to unload.
  */
  async unloadModule (module: number): Promise<Module> {
    const query: PAPacket = UnloadModule.query(this.requestId(), module)
    return await this.sendRequest(query)
  }

  // Private methods
  private onReadable (): void {
    // Don't read if we don't have at least 10 bytes
    if (this.socket.readableLength < 10) {
      console.log('Malformed packet!')
      const malformed: Buffer = this.socket.read(this.socket.readableLength)
      console.log(malformed)
      return
    }

    const chunk: Buffer = this.socket.read(this.socket.readableLength)

    // Discard previous chunks if we get a new packet header and stil have pending chunks
    // This should not happen really, if it does it's probably a bug
    if (PAPacket.isChunkHeader(chunk) && this.chunks.length > 0) {
      console.log('WARNING: discarding chunks, this might be a bug...')
      this.chunks = []
    }

    // Save new chunk
    this.chunks.push(chunk)

    // Once we get the entire package fulfill the request promise
    // Chunks can contain more than one packet so we loop until we are done
    while (PAPacket.isValidPacket(this.chunks)) {
      const packet: PAPacket = new PAPacket(Buffer.concat(this.chunks))
      this.resolveRequest(packet)
      this.chunks = PAPacket.getChunksSize(this.chunks) === packet.packet.length ? [] : [Buffer.concat(this.chunks).subarray(packet.packet.length)]
    }
  }

  private requestId (): number {
    this.lastRequestId = (this.lastRequestId + 1) & PA_MAX_REQUEST_ID
    return this.lastRequestId
  }

  private async sendRequest (query: PAPacket): Promise<any> {
    const request: PARequest = new PARequest(this.lastRequestId, query)
    this.requests.push(request)

    if (!this.connected && query.command.value !== PA_NATIVE_COMMAND_NAMES.PA_COMMAND_AUTH.toString().charCodeAt(0)) {
      return this.rejectRequest(request, new Error('No connection to PulseAudio.'))
    }

    this.socket.write(request.query.write())
    return await request.promise
  }

  private resolveRequest (reply: PAPacket): void {
    let request: PARequest | undefined
    let event: PAEvent | undefined
    switch (reply.command.value) {
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_ERROR:
        request = this.requests.find(r => r.id === reply.requestId.value)
        request?.resolve({ success: false, error: PAError[reply.tags[0].value] })
        this.requests = this.requests.filter(r => r.id !== reply.requestId.value)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_REPLY:
        request = this.requests.find(r => r.id === reply.requestId.value)
        request?.resolve(this.parseReply(reply, request.query.command))
        this.requests = this.requests.filter(r => r.id !== reply.requestId.value)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_SUBSCRIBE_EVENT:
        event = this.parseEvent(reply)
        this.emit(event.category, event)
        this.emit('all', event)
        break
      default:
        throw new Error(`Reply type ${reply.command.value} not supported. Please report issue.`)
    }
  }

  private rejectRequest (request: PARequest, error: Error): void {
    request.reject(error)
    this.requests = this.requests.filter(r => r.id !== request.id)
  }

  private parseReply (reply: PAPacket, query: PATag<any>): any {
    let retObj: any = {}

    switch (query.value) {
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_AUTH:
        retObj = Authenticate.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_SERVER_INFO:
        retObj = GetServerInfo.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_SUBSCRIBE:
        retObj = Subscribe.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_SET_CLIENT_NAME:
        retObj = SetClientName.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_SINK_INFO_LIST:
        retObj = GetSinkList.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_SINK_INFO:
        retObj = GetSink.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_SET_SINK_VOLUME:
        retObj = SetSinkVolume.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_SINK_INPUT_INFO:
        retObj = GetSinkInput.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_SINK_INPUT_INFO_LIST:
        retObj = GetSinkInputList.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_MOVE_SINK_INPUT:
        retObj = MoveSinkInput.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_SOURCE_INFO:
        retObj = GetSource.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_SOURCE_INFO_LIST:
        retObj = GetSourceList.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_SET_SOURCE_VOLUME:
        retObj = SetSourceVolume.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_SOURCE_OUTPUT_INFO:
        retObj = GetSourceOutput.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_SOURCE_OUTPUT_INFO_LIST:
        retObj = GetSourceOutputList.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_MOVE_SOURCE_OUTPUT:
        retObj = MoveSourceOutput.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_MODULE_INFO:
        retObj = GetModule.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_GET_MODULE_INFO_LIST:
        retObj = GetModuleList.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_LOAD_MODULE:
        retObj = LoadModule.reply(reply, this.protocol)
        break
      case PA_NATIVE_COMMAND_NAMES.PA_COMMAND_UNLOAD_MODULE:
        retObj = UnloadModule.reply(reply, this.protocol)
        break
      default:
        throw new Error(`Command ${query.value as string} not supported. Please report issue.`)
    }
    return retObj
  }

  private parseEvent (packet: PAPacket): PAEvent {
    const details: number = packet.tags[0].value
    const index: number = packet.tags[1].value

    // Parse category
    let category: string = ''
    switch (details & PASubscriptionEventType.FACILITY_MASK) {
      case PASubscriptionEventType.SINK:
        category = 'sink'
        break
      case PASubscriptionEventType.SOURCE:
        category = 'source'
        break
      case PASubscriptionEventType.SINK_INPUT:
        category = 'sinkInput'
        break
      case PASubscriptionEventType.SOURCE_OUTPUT:
        category = 'sourceOutput'
        break
      case PASubscriptionEventType.MODULE:
        category = 'module'
        break
      case PASubscriptionEventType.CLIENT:
        category = 'client'
        break
      case PASubscriptionEventType.SAMPLE_CACHE:
        category = 'sampleCache'
        break
      case PASubscriptionEventType.SERVER:
        category = 'server'
        break
      case PASubscriptionEventType.AUTOLOAD:
        category = 'autoload'
        break
      case PASubscriptionEventType.CARD:
        category = 'card'
        break
      default:
        throw new Error(`Details type ${details} not supported. Please report issue.`)
    }

    // Parse event type
    let type: string = ''
    switch (details & PASubscriptionEventType.TYPE_MASK) {
      case PASubscriptionEventType.NEW:
        type = 'new'
        break
      case PASubscriptionEventType.CHANGE:
        type = 'change'
        break
      case PASubscriptionEventType.REMOVE:
        type = 'remove'
        break
      default:
        throw new Error(`Event type ${details} not supported. Please report issue.`)
    }

    return {
      index,
      category,
      type
    }
  }

  private parseAddress (address: string): void {
    // reference = tcp:pulseaudio:4317
    if (address.startsWith('tcp:')) {
      const split: string[] = address.split(':')
      this.address = {
        type: 'tcp',
        port: parseInt(split[2] ?? '4317'),
        host: split[1]
      }
    // reference = unix:/path/to/socket
    } else if (address.startsWith('unix:')) {
      const split: string[] = address.split(':')
      this.address = {
        type: 'unix',
        path: split[1]
      }
    } else if (address.includes(':')) {
      const split: string[] = address.split(':')
      this.address = {
        type: 'tcp',
        port: parseInt(split[1] ?? '4317'),
        host: split[0]
      }
    } else {
      throw new Error('Unrecognized server address format. Please use "tcp:host:port", "unix:/path/to/socket" or "host:port".')
    }
  }

  private parseCookie (cookiePath: string): void {
    try {
      this.cookie = Buffer.from(readFileSync(cookiePath, 'hex'), 'hex')
    } catch (error) {
      console.log('Error reading cookie file, might not be able to authenticate.')
      console.log(error)
    }
  }
}
